pages/agent/v3/hooks.md

```diff
@@ -137,29 +136,34 @@ they are run as part of each job:
 
 ### Creating job lifecycle hooks
 
-Job lifecycle hooks are sourced for every job an agent accepts.
-Use job lifecycle hooks to prepare for jobs, override default behavior, or clean up after jobs that have finished.
-For example, use the `environment` hook to set a job's environment variables or the `pre-exit` hook to delete temporary files and remove containers.
-If your hook is related to the startup or shutdown of the agent, consider [agent lifecycle hooks](#job-lifecycle-hooks) for those tasks instead.
+Job lifecycle hooks are sourced for every job an agent accepts.  Use job lifecycle hooks to prepare for jobs, override default behavior, or clean up after jobs that have finished.  For example, use the `environment` hook to set a job's environment variables or the `pre-exit` hook to delete temporary files and remove containers.  If your hook is related to the startup or shutdown of the agent, consider [agent lifecycle hooks](#job-lifecycle-hooks) for those tasks instead.
 
 Job lifecycle hooks have access to all the standard [Buildkite environment variables](/docs/pipelines/environment-variables).
 
-Job lifecycle hooks are copied to `$TMPDIR` directory and *sourced* by the agent's default shell.
-This has a few implications:
+Job lifecycle hooks are copied to `$TMPDIR` directory and *sourced* by the agent's default shell. This has a few implications:
+
+#### \<aside\>
+#### 📝 A note on sourcing for shellscripting nerds
+
+We use the word "sourcing" in this article, but it's not strictly speaking true; we're lying by omission a little bit. Instead, the agent uses a process called ["the scriptwrapper"](https://github.com/buildkite/agent/blob/1a5f05029cc363a984188c441f938dd316dedd16/hook/scriptwrapper.go).
+
+This process notes down the environment variables prior to a hook run, sources that hook, and then compares the environment variables after the hook run to the environment variables before the hook run.
+
+Any environment variables that were added, changed or removed are then exported to the subsequent phases and hooks. Functionally, this is very similar to how `source` would work, but it's not quite the same, and it means that if you're relying on some very specific pieces of shellscripting arcana, you might find that things don't work quite as you expect.
+
+The reason that we do this is that there's no shared bash environment between two different hooks on the same job; functionally, each hook is run in its own shell, which are orchestrated through the Agent's Go code. This means that if you set an environment variable in one hook, it wouldn't be available in the next hook without this scriptwrapper process.
+
+#### \<\/aside\>
```

Comment:
@mbelton-buildkite i would like this information to present in the docs - it's useful context for how hooks actually work, but i can't think of a nice way to present it without breaking the flow of the docs - can you think of a good way to format this / place to put it?

Reply:
Interesting! Okay, so this is how I'm thinking about it:

- Who needs this information? (Is it everyone using hooks? Is it most people? Is it so special-interest niche that it's almost no one?)
- When do you need this information? (Is it more when you're learning about the feature or when you've run into a problem you're trying to solve?)
- How important is this information when you need it? (How badly will someone be blocked/confused without this?)
- What other info does it relate to? I see it's only referenced in the "Hook scopes" section, so can it go there?

My initial thoughts are:

1. If most people need this info and it's important they find it, it should have its own heading and be included on the page. Near where you reference it if it's fairly important, or at the bottom if it's not.
1. If it's mainly additional context that some people would be interested in, but it's probably not blocking someone, perhaps we could include it in an expand. I haven't seen that used in our docs, but I think it can be useful in cases like this.

I feel like option 2 is a good fit for this, but I might be missing its importance.

I know that's not a clear answer, but you'll know this better than me. Happy to talk it through if that would be easier. 😊

Reply:
yeah, i think i'm an option 2 kinda person too - it's pretty unlikely that this information is like, actually useful to someone. imo tho it's useful context, and given that calling what we do sourcing is kind of a lie, worth noting.

how would one put an expand in? using a `<details>` block like in reggo markdown?

Reply:
I had a quick check and the standard `<details>` should work:

```
<details>
  <summary>Epcot Center</summary>
  <p>Epcot is a theme park at Walt Disney World Resort featuring exciting attractions, international pavilions, award-winning fireworks and seasonal special events.</p>
</details>
```

I believe it does mean you'll need to use HTML for formatting inside it.

Reply:
ah, useful to know! i'll knock something out :)